Master add ad7768 1 support

stefpopa opened the pull request:
This PR adds basic support for AD7768-1 ADC. 

The devices is configured in fast mode, with 32 kSPS and leaves the default sinc5 filter.

Two data conversion modes are made available. When data is retrieved by using the read_raw attribute, one shot single conversion mode is set. The continuous conversion mode is enabled when the triggered buffer mechanism is used. To assure correct data retrieval, the driver waits for the interrupt triggered by the low to high transition of the DRDY pin.

Another PR will add additional features such as the configuration of sampling frequency and bandwidth.

stefpopa commented:
Will merge later today, if there are no objections.

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,468 @@
+		if (scale_uv < 0)
+			return scale_uv;
+
+		*val = scale_uv * 2 / 1000;
```
nitpick: i'd add a parenthesis here -> *val = (scale_uv * 2) / 1000;
i typically never know [or care about] order of evaluation of arithmetic operations [if parantheses are added ]

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,468 @@
+		return -ENODEV;
+	}
+
+	/* Perform a soft reset at startup by writing these magic numbers */
```
curios: are these numbers from the datasheet ? maybe it's worth specifying this [for upstream review mostly] 
[ i admit i was lazy in looking up the datasheet ]

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,468 @@
+	if (iio_buffer_enabled(indio_dev))
+		iio_trigger_poll_chained(st->trig);
+	else
+		complete(&st->completion);
```
this only gets inited once per driver init;
curios [i'm not very familiar with completion structs & and their behavior], but would this need to be reinited [via reinit_completion] when enabling/disabling some buffer mode ? or is this sufficient ?

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,468 @@
+		 * Any SPI configuration of the AD7768-1 can only be
+		 * performed in continuous conversion mode.
+		 */
+		ret = ad7768_set_mode(st, AD7768_CONTINUOUS);
```
when the device resets, is it in continuous mode ? or does this need to be set in the setup/probe phase ?

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,468 @@
+	if (ret)
+		return ret;
+
+	ret = ad7768_spi_reg_write(st, AD7768_REG_SYNC_RESET, 0x82);
```
maybe in a later change, a reset GPIO can be added to this, as a first reset mechanism

stefpopa commented:
changes in v3:
* added parenthesis: `*val = (scale_uv * 2) / 1000;`
* added a more verbose comment when the software reset is performed.
* called `reinit_completion()` every time a new single conversion is performed.

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,475 @@
+	struct ad7768_state *st = iio_priv(indio_dev);
+	int ret;
+
+	reinit_completion(&st->completion);
```
i am not sure this is the best place to call `reinit_completion`; 
typically, the way this seems to work is
```
1. call `reinit_completion`
2. start a procedure
3. call `wait_for_completion_timeout`
```

maybe in this case, `reinit_completion` should be called before ` ad7768_set_mode(st, AD7768_ONE_SHOT);`
in which case, it may be a good idea to move `ad7768_set_mode(st, AD7768_ONE_SHOT)` here

stefpopa commented:
changes in v4:
* called `reinit_completion ` before setting the one shot conversion mode.

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,475 @@
+			return ret;
+
+		ret = ad7768_scan_direct(indio_dev);
+		if (ret < 0)
```
nitpick: particularly here it's a bit neater to do
```
if (ret == 0)
   *val = ret;
```
than use a `goto` label

commodo replied:
that is, unless you are adding something later between this code and the label

stefpopa commented:
changes in v5:
* removed the `goto `label from `read_raw()`

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,473 @@
+	 * To exit continuous read mode, perform a single read of the ADC_DATA
+	 * reg (0x2C), which allows further configuration of the device.
+	 */
+	ret = ad7768_spi_reg_read(st, AD7768_REG_ADC_DATA, 3);
```
Will you add more stuff here ?
If not, then you could just return directly

stefpopa replied:
The problem is that `ad7768_spi_reg_read()` returns the `ADC_DATA `reg value or a negative value in case of error. I am not sure if the buffer_ops will like this.

commodo replied:
ack, i see now
thanks for the clarification

commodo commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -0,0 +1,473 @@
+	unsigned int shift;
+	int ret;
+
+	shift = 32 - 8 * len;
```
nitpick: parentheses `32 - (8 * len)`

stefpopa replied:
will fix this

commodo commented:
rest looks good

stefpopa commented:
changes in v6:
* added parentheses for the expression: `32 - (8 * len)`
